File: backend/app/utils/foundation_groups.py
"""Which foundation class group somebody is in.

Recorded, not derived. This used to be a function of the date - the 1st-15th
was Group 1 and the 16th onward Group 2 - on the reasoning that which sitting
somebody came through is a fact about when, not a decision anybody makes. That
turned out to be wrong twice over: there are three groups, not two halves of a
month, and students are moved between them. A rule that reads a date can be
told about neither.

So the group is asked for on the Induction Call Form, stored on the entry,
carried onto the Foundation lead when the two are linked, and editable on the
boards that show it. Every move is kept beside it (FoundationGroupMove in
app/models/foundation_group.py), because "moved to Group 2 from Group 1" is the
part people actually need to read.

What is left here is the small vocabulary the rest of the app shares: the
default set of groups, how a group is written down, and how the label a form
sends is read back into the number that is stored.
"""
from app.exceptions.base import BadRequestError

# The groups the Induction Call Form offers out of the box. Only a default:
# the real list is the `group` field's options in the form config, which an
# admin edits, so nothing here refuses a fourth group the day one is added.
# This is what the boards' filter dropdowns fall back to.
FOUNDATION_GROUPS = (1, 2, 3)

# An upper bound on what can be stored, not on how many groups there can be.
# It exists so a date or a phone number typed into the box is refused as the
# mistake it is rather than filed as group 2026.
MAX_FOUNDATION_GROUP = 99


def foundation_group_label(group: int | None) -> str:
    """"Group 2", or "" for a row that has no group yet."""
    return f"Group {group}" if group else ""


def parse_foundation_group(value: str | int | None) -> int | None:
    """The number behind whatever the form sent.

    Takes "Group 2", "group2", "2" and 2 alike, because the form sends the
    label an admin typed into the options box and there is no reason to make
    the two agree on punctuation. Blank means "not said", which is a legitimate
    answer while the field is optional.

    Anything else raises rather than being dropped: a group that silently
    failed to save would put the student in no class at all, and the person who
    picked one would have no way of knowing.
    """
    if value is None:
        return None
    if isinstance(value, bool):  # bool is an int; "group True" is not a group.
        raise BadRequestError("Pick a group, for example Group 1.")
    if isinstance(value, int):
        number = value
    else:
        digits = "".join(character for character in str(value) if character.isdigit())
        if not str(value).strip():
            return None
        if not digits:
            raise BadRequestError(f'"{value}" doesn\'t name a group. Use Group 1, Group 2 or Group 3.')
        # The whole run of digits, not a prefix of it: "2026" has to fail
        # the bound below as the year it is, rather than pass as group 20.
        number = int(digits) if len(digits) <= 4 else MAX_FOUNDATION_GROUP + 1
    if not 1 <= number <= MAX_FOUNDATION_GROUP:
        raise BadRequestError(f"Group {number} isn't a group. Use Group 1, Group 2 or Group 3.")
    return number

File: frontend/src/constants/foundationGroups.js
// The foundation class groups a student can be put in.
//
// Recorded, not derived. This used to read a date - the 1st-15th of the month
// was Group 1, the 16th onward Group 2 - which meant the group was whatever
// the calendar said and nobody could correct it. There are three classes, not
// two halves of a month, and students are moved between them; a rule that
// reads a date can express neither.
//
// The backend owns the field now (app/utils/foundation_groups.py). This file
// is only how the boards say it: the fallback option list, and the colours.
export const FOUNDATION_GROUPS = [1, 2, 3]

// The filter dropdowns take string values, so these are strings while the API
// field is a number.
//
// A fixed list rather than options read off the data, for the same reason the
// section filter has one: "nobody is in Group 3 this month" is an answer the
// filter should be able to give rather than an option it quietly drops.
export const FOUNDATION_GROUP_OPTIONS = FOUNDATION_GROUPS.map((group) => ({
  value: String(group),
  label: `Group ${group}`,
}))

// What the editable cells offer. The same labels the Induction Call Form's
// Group dropdown is seeded with, so a group picked on the board and a group
// picked on the form are written the same way.
export const FOUNDATION_GROUP_LABELS = FOUNDATION_GROUPS.map((group) => `Group ${group}`)

// One hue per group so a column of them is scannable - which class a row
// belongs to is the whole point of the column, and three shades of one colour
// would need reading rather than seeing.
const TONES = { 1: 'blue', 2: 'violet', 3: 'teal' }

export const foundationGroupTone = (group) => TONES[group] ?? 'slate'
export const foundationGroupLabel = (group) => (group ? `Group ${group}` : null)

// "Group 2" -> 2. The cells store and send a number; the dropdowns they are
// built from deal in labels, because that is what an admin types into the
// form's option list. Anything with no digits in it is not a group.
export function foundationGroupValue(label) {
  if (!label) return null
  const digits = String(label).match(/\d+/)
  return digits ? Number(digits[0]) : null
}

// The most recent move, or null for a student who has never been moved.
//
// A move *between* groups: being put into one in the first place is in the
// history too, and is not something the board should announce - every student
// was put into a group once, and a column that said so about all of them would
// be a column of noise.
//
// A change recorded as direct ("belongs in Group 3", not "moved to it") wipes
// the note: the latest change is what the row describes, and it says there
// was no move.
export function lastGroupMove(history) {
  const moves = (history ?? []).filter((move) => move.from_group)
  const last = moves[moves.length - 1]
  return last && !last.direct ? last : null
}
